File: src/bundler_jsc/PluginRunner.zig
//! Runtime plugin host (JS-side `Bun.plugin()` resolve hooks). Moved from
//! `bundler/transpiler.zig` so `bundler/` is free of `JSValue`/`JSGlobalObject`.

pub const MacroJSCtx = jsc.JSValue;
pub const default_macro_js_value = jsc.JSValue.zero;

pub const PluginRunner = struct {
    global_object: *jsc.JSGlobalObject,
    allocator: std.mem.Allocator,

    pub fn extractNamespace(specifier: string) string {
        const colon = strings.indexOfChar(specifier, ':') orelse return "";
        if (Environment.isWindows and
            colon == 1 and
            specifier.len > 3 and
            bun.path.isSepAny(specifier[2]) and
            ((specifier[0] > 'a' and specifier[0] < 'z') or (specifier[0] > 'A' and specifier[0] < 'Z')))
            return "";
        return specifier[0..colon];
    }

    pub fn couldBePlugin(specifier: string) bool {
        if (strings.lastIndexOfChar(specifier, '.')) |last_dor| {
            const ext = specifier[last_dor + 1 ..];
            // '.' followed by either a letter or a non-ascii character
            // maybe there are non-ascii file extensions?
            // we mostly want to cheaply rule out "../" and ".." and "./"
            if (ext.len > 0 and ((ext[0] >= 'a' and ext[0] <= 'z') or (ext[0] >= 'A' and ext[0] <= 'Z') or ext[0] > 127))
                return true;
        }
        return (!std.fs.path.isAbsolute(specifier) and strings.containsChar(specifier, ':'));
    }

    pub fn onResolve(
        this: *PluginRunner,
        specifier: []const u8,
        importer: []const u8,
        log: *logger.Log,
        loc: logger.Loc,
        target: jsc.JSGlobalObject.BunPluginTarget,
    ) bun.JSError!?Fs.Path {
        var global = this.global_object;
        const namespace_slice = extractNamespace(specifier);
        const namespace = if (namespace_slice.len > 0 and !strings.eqlComptime(namespace_slice, "file"))
            bun.String.init(namespace_slice)
        else
            bun.String.empty;
        const on_resolve_plugin = try global.runOnResolvePlugins(
            namespace,
            bun.String.init(specifier).substring(if (namespace.length() > 0) namespace.length() + 1 else 0),
            bun.String.init(importer),
            target,
        ) orelse return null;
        const path_value = try on_resolve_plugin.get(global, "path") orelse return null;
        if (path_value.isEmptyOrUndefinedOrNull()) return null;
        if (!path_value.isString()) {
            log.addError(null, loc, "Expected \"path\" to be a string") catch unreachable;
            return null;
        }

        const file_path = try path_value.toBunString(global);
        defer file_path.deref();

        if (file_path.length() == 0) {
            log.addError(
                null,
                loc,
                "Expected \"path\" to be a non-empty string in onResolve plugin",
            ) catch unreachable;
            return null;
        } else if
        // TODO: validate this better
        (file_path.eqlComptime(".") or
            file_path.eqlComptime("..") or
            file_path.eqlComptime("...") or
            file_path.eqlComptime(" "))
        {
            log.addError(
                null,
                loc,
                "Invalid file path from onResolve plugin",
            ) catch unreachable;
            return null;
        }
        var static_namespace = true;
        const user_namespace: bun.String = brk: {
            if (try on_resolve_plugin.get(global, "namespace")) |namespace_value| {
                if (!namespace_value.isString()) {
                    log.addError(null, loc, "Expected \"namespace\" to be a string") catch unreachable;
                    return null;
                }

                const namespace_str = try namespace_value.toBunString(global);
                if (namespace_str.length() == 0) {
                    namespace_str.deref();
                    break :brk bun.String.init("file");
                }

                if (namespace_str.eqlComptime("file")) {
                    namespace_str.deref();
                    break :brk bun.String.init("file");
                }

                if (namespace_str.eqlComptime("bun")) {
                    namespace_str.deref();
                    break :brk bun.String.init("bun");
                }

                if (namespace_str.eqlComptime("node")) {
                    namespace_str.deref();
                    break :brk bun.String.init("node");
                }

                static_namespace = false;

                break :brk namespace_str;
            }

            break :brk bun.String.init("file");
        };
        defer user_namespace.deref();

        if (static_namespace) {
            return Fs.Path.initWithNamespace(
                std.fmt.allocPrint(this.allocator, "{f}", .{file_path}) catch unreachable,
                user_namespace.byteSlice(),
            );
        } else {
            return Fs.Path.initWithNamespace(
                std.fmt.allocPrint(this.allocator, "{f}", .{file_path}) catch unreachable,
                std.fmt.allocPrint(this.allocator, "{f}", .{user_namespace}) catch unreachable,
            );
        }
    }

    pub fn onResolveJSC(this: *const PluginRunner, namespace: bun.String, specifier: bun.String, importer: bun.String, target: jsc.JSGlobalObject.BunPluginTarget) bun.JSError!?jsc.ErrorableString {
        var global = this.global_object;
        const on_resolve_plugin = try global.runOnResolvePlugins(
            if (namespace.length() > 0 and !namespace.eqlComptime("file"))
                namespace
            else
                bun.String.static(""),
            specifier,
            importer,
            target,
        ) orelse return null;
        if (!on_resolve_plugin.isObject()) return null;
        const path_value = try on_resolve_plugin.get(global, "path") orelse return null;
        if (path_value.isEmptyOrUndefinedOrNull()) return null;
        if (!path_value.isString()) {
            return jsc.ErrorableString.err(
                error.JSErrorObject,
                bun.String.static("Expected \"path\" to be a string in onResolve plugin").toErrorInstance(this.global_object),
            );
        }

        const file_path = try path_value.toBunString(global);

        if (file_path.length() == 0) {
            return jsc.ErrorableString.err(
                error.JSErrorObject,
                bun.String.static("Expected \"path\" to be a non-empty string in onResolve plugin").toErrorInstance(this.global_object),
            );
        } else if
        // TODO: validate this better
        (file_path.eqlComptime(".") or
            file_path.eqlComptime("..") or
            file_path.eqlComptime("...") or
            file_path.eqlComptime(" "))
        {
            return jsc.ErrorableString.err(
                error.JSErrorObject,
                bun.String.static("\"path\" is invalid in onResolve plugin").toErrorInstance(this.global_object),
            );
        }
        var static_namespace = true;
        const user_namespace: bun.String = brk: {
            if (try on_resolve_plugin.get(global, "namespace")) |namespace_value| {
                if (!namespace_value.isString()) {
                    return jsc.ErrorableString.err(
                        error.JSErrorObject,
                        bun.String.static("Expected \"namespace\" to be a string").toErrorInstance(this.global_object),
                    );
                }

                const namespace_str = try namespace_value.toBunString(global);
                if (namespace_str.length() == 0) {
                    break :brk bun.String.static("file");
                }

                if (namespace_str.eqlComptime("file")) {
                    defer namespace_str.deref();
                    break :brk bun.String.static("file");
                }

                if (namespace_str.eqlComptime("bun")) {
                    defer namespace_str.deref();
                    break :brk bun.String.static("bun");
                }

                if (namespace_str.eqlComptime("node")) {
                    defer namespace_str.deref();
                    break :brk bun.String.static("node");
                }

                static_namespace = false;

                break :brk namespace_str;
            }

            break :brk bun.String.static("file");
        };
        defer user_namespace.deref();

        // Our super slow way of cloning the string into memory owned by jsc
        const combined_string = std.fmt.allocPrint(this.allocator, "{f}:{f}", .{ user_namespace, file_path }) catch unreachable;
        var out_ = bun.String.init(combined_string);
        defer out_.deref();
        const jsval = out_.toJS(this.global_object) catch |err| {
            this.allocator.free(combined_string);
            return jsc.ErrorableString.err(err, this.global_object.tryTakeException() orelse .js_undefined);
        };
        const out = jsval.toBunString(this.global_object) catch |err| {
            this.allocator.free(combined_string);
            return jsc.ErrorableString.err(err, this.global_object.tryTakeException() orelse .js_undefined);
        };
        this.allocator.free(combined_string);
        return jsc.ErrorableString.ok(out);
    }
};

const string = []const u8;

const std = @import("std");

const bun = @import("bun");
const Environment = bun.Environment;
const Fs = bun.fs;
const jsc = bun.jsc;
const logger = bun.logger;
const strings = bun.strings;

File: src/jsc/PluginRunner.rs
//! The concrete JSC-aware plugin-resolve hook.
//!
//! LAYERING: the static byte helpers (`extract_namespace` / `could_be_plugin`)
//! live in `bun_bundler::transpiler::PluginRunner` (JSC-free, lowest tier).
//! The stateful struct lives here because its only field is a typed
//! `*mut JSGlobalObject`. `bun_bundler::Linker` references it through
//! `*mut dyn PluginResolver`, so the linker stays JSC-free without
//! type-erasing to `*mut c_void` or duplicating the body behind a fn-ptr.

use std::io::Write as _;

use bun_bundler::transpiler::{BunPluginTarget, PluginResolver};
use bun_core::{OwnedString, String as BunString};
use bun_paths::fs::Path as FsPath;

use crate::JSGlobalObject;
use bun_ptr::BackRef;

pub struct PluginRunner {
    pub global_object: BackRef<JSGlobalObject>,
}

// Re-export the JSC-free static helpers so callers in this crate can keep
// writing `PluginRunner::could_be_plugin(...)` without naming `bun_bundler`.
impl PluginRunner {
    /// Borrow the JS global stored by `Bun__onDidAppendPlugin`.
    ///
    /// SAFETY (invariant): `global_object` is the live `*mut JSGlobalObject`
    /// installed by `Bun__onDidAppendPlugin`; the VM (and its global) outlives
    /// every `Linker::link` call that reaches plugin hooks. Never null.
    #[inline]
    fn global(&self) -> &JSGlobalObject {
        self.global_object.get()
    }

    #[inline]
    pub fn extract_namespace(specifier: &[u8]) -> &[u8] {
        bun_bundler::transpiler::PluginRunner::extract_namespace(specifier)
    }
    #[inline]
    pub fn could_be_plugin(specifier: &[u8]) -> bool {
        bun_bundler::transpiler::PluginRunner::could_be_plugin(specifier)
    }
}

impl PluginResolver for PluginRunner {
    fn on_resolve(
        &self,
        specifier: &[u8],
        importer: &[u8],
        log: &mut bun_ast::Log,
        loc: bun_ast::Loc,
        target: BunPluginTarget,
    ) -> bun_bundler::Result<Option<FsPath<'static>>> {
        let global = self.global();
        let js_err = |e: crate::JsError| {
            bun_bundler::Error::Js(match e {
                crate::JsError::Thrown => bun_core::JsError::Thrown,
                crate::JsError::OutOfMemory => bun_core::JsError::OutOfMemory,
                crate::JsError::Terminated => bun_core::JsError::Terminated,
            })
        };

        let namespace_slice = Self::extract_namespace(specifier);
        let namespace = if !namespace_slice.is_empty() && namespace_slice != b"file" {
            BunString::init(namespace_slice)
        } else {
            BunString::empty()
        };
        let Some(on_resolve_plugin) = global
            .run_on_resolve_plugins(
                namespace,
                BunString::init(specifier).substring(if namespace.length() > 0 {
                    namespace.length() + 1
                } else {
                    0
                }),
                BunString::init(importer),
                target,
            )
            .map_err(js_err)?
        else {
            return Ok(None);
        };
        let Some(path_value) = on_resolve_plugin.get(global, "path").map_err(js_err)? else {
            return Ok(None);
        };
        if path_value.is_empty_or_undefined_or_null() {
            return Ok(None);
        }
        if !path_value.is_string() {
            log.add_error(None, loc, b"Expected \"path\" to be a string");
            return Ok(None);
        }

        // `bun_core::String`
        // is `Copy` (no `Drop`), so RAII-wrap the +1 WTF ref across every
        // remaining `?` / early-return.
        let file_path = OwnedString::new(path_value.to_bun_string(global).map_err(js_err)?);

        if file_path.length() == 0 {
            log.add_error(
                None,
                loc,
                b"Expected \"path\" to be a non-empty string in onResolve plugin",
            );
            return Ok(None);
        } else if
        // TODO: validate this better
        file_path.eql_comptime(b".")
            || file_path.eql_comptime(b"..")
            || file_path.eql_comptime(b"...")
            || file_path.eql_comptime(b" ")
        {
            log.add_error(None, loc, b"Invalid file path from onResolve plugin");
            return Ok(None);
        }
        let mut static_namespace = true;
        let user_namespace: BunString = 'brk: {
            if let Some(namespace_value) =
                on_resolve_plugin.get(global, "namespace").map_err(js_err)?
            {
                if !namespace_value.is_string() {
                    log.add_error(None, loc, b"Expected \"namespace\" to be a string");
                    return Ok(None);
                }

                let namespace_str = namespace_value.to_bun_string(global).map_err(js_err)?;
                if namespace_str.length() == 0 {
                    namespace_str.deref();
                    break 'brk BunString::init(b"file");
                }

                if namespace_str.eql_comptime(b"file") {
                    namespace_str.deref();
                    break 'brk BunString::init(b"file");
                }

                if namespace_str.eql_comptime(b"bun") {
                    namespace_str.deref();
                    break 'brk BunString::init(b"bun");
                }

                if namespace_str.eql_comptime(b"node") {
                    namespace_str.deref();
                    break 'brk BunString::init(b"node");
                }

                static_namespace = false;

                break 'brk namespace_str;
            }

            break 'brk BunString::init(b"file");
        };
        let user_namespace = OwnedString::new(user_namespace);

        // `FsPath<'static>` borrows, so the formatted buffer is leaked to
        // model the caller-owns-forever contract (these are
        // never freed — the linker arena owns them for the build).
        // PERF: profile if hot.
        let mut path_buf: Vec<u8> = Vec::new();
        write!(&mut path_buf, "{}", file_path).expect("unreachable");
        let path_static: &'static [u8] = path_buf.leak();

        if static_namespace {
            // `byte_slice()` borrows `&self`; re-match to recover the
            // `'static` literal so the result typechecks as `FsPath<'static>`
            // without an extra alloc.
            let ns: &'static [u8] = if user_namespace.eql_comptime(b"bun") {
                b"bun"
            } else if user_namespace.eql_comptime(b"node") {
                b"node"
            } else {
                b"file"
            };
            Ok(Some(FsPath::init_with_namespace(path_static, ns)))
        } else {
            let mut ns_buf: Vec<u8> = Vec::new();
            write!(&mut ns_buf, "{}", user_namespace).expect("unreachable");
            let ns_static: &'static [u8] = ns_buf.leak();
            Ok(Some(FsPath::init_with_namespace(path_static, ns_static)))
        }
    }
}
